pkg/workflow/controller/handlers/workflowrun/handler.go

```diff
@@ -84,11 +104,45 @@ func (h *Handler) ObjectUpdated(obj interface{}) {
 	// the GC queue.
 	h.GCProcessor.Add(originWfr)
 
-	// If the WorkflowRun has already been terminated(Completed, Error, Cancel) or waiting for external events, skip it.
+	// If the WorkflowRun has already been terminated(Completed, Error, Cancel), send notifications if necessary,
+	// otherwise directly skip it.
 	if originWfr.Status.Overall.Status == v1alpha1.StatusCompleted ||
 		originWfr.Status.Overall.Status == v1alpha1.StatusError ||
-		originWfr.Status.Overall.Status == v1alpha1.StatusWaiting ||
 		originWfr.Status.Overall.Status == v1alpha1.StatusCancelled {
+		// No need to send notifications for workflowruns finished before workflow controller starts.
+		if originWfr.Status.Overall.LastTransitionTime.Before(controllerStartTime) {
+			return
+		}
+
+		// Send notifications with workflowrun.
+		bodyBytes, err := json.Marshal(originWfr)
```

Comment:
I'd like a field in status of wfr to mark whether notification been sent and the send result(send successfully, or failed). This should be a status of wfr.

And I think there is no need to send the whole `wfr`. We need define the notification request as need.

Reply:
> I'd like a field in status of wfr to mark whether notification been sent and the send result(send successfully, or failed). This should be a status of wfr.

This can be added as an enhancement later.

> And I think there is no need to send the whole `wfr`. We need define the notification request as need.

Sending the whole `wfr` to keep the extensibility, implementers can get what they need from `wfr`.

Reply:
> Sending the whole wfr to keep the extensibility, implementers can get what they need from wfr.

I hold different opinions against this, couple the API spec to the definition of wfr is not a good pattern. wfr is easier to be modifed (add new fields, rename fields), but API spec should consider compatibility.

Reply:
wfr is a part of API spec as there is CRUD API for it, we should also consider its compatibility.

Reply:
Notification and wfr CRUD are two different API sets, they should not bound together. If we have to break the compatibility of wfr one day, it wont affect notifications.

Reply:
Have discussed offline, and agreed with:

Notification should include basic informations such as tenant, project, etc, but workflow engine does not aware about these concepts. Workflow engine can not provide these necessary informations, so we directly send `wfr` in notification and let implementers get what they need from it.